src/saltext/vmware/modules/datastore.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License: Apache-2.0
+import logging
+
+import salt.exceptions
+import saltext.vmware.utils.common as utils_common
+import saltext.vmware.utils.esxi as utils_esxi
+import saltext.vmware.utils.vmware as utils_vmware
+from saltext.vmware.utils.connect import get_service_instance
+
+log = logging.getLogger(__name__)
+
+try:
+    from pyVmomi import vmodl, vim, VmomiSupport
+
+    HAS_PYVMOMI = True
+except ImportError:
+    HAS_PYVMOMI = False
+
+
+__virtualname__ = "vmware_datastore"
+
+
+def __virtual__():
+    if not HAS_PYVMOMI:
+        return False, "Unable to import pyVmomi module."
+    return __virtualname__
+
+
+def _find_filtered_object(service_instance, datacenter_name=None, cluster_name=None, host_name=None):
+    """
+    finds zero or one matching objects: plug in almost any combination of datacenter, cluster, and/or host name
+    *if cluster_name is passed, datacenter_name must also be passed
+
+    The most specific object will be returned.
+
+    If the combination of parameters has the *potential* to match multiple objects, an exception is raised.
+
+    At least one of these parameters must be set.
+
+    service_instance
+        The Service Instance Object from which to obtain cluster.
+
+    datacenter_name
+        (Optional) Datacenter name to filter by.
+
+    cluster_name
+        (Optional) Exact cluster name to filter by. If used, datacenter_name is required.
+
+    host_name
+        (Optional) Exact host name name to filter by.
+    """
+    if host_name:
+        objects = utils_esxi.get_hosts(service_instance,
+                                       datacenter_name=datacenter_name,
+                                       cluster_name=cluster_name,
+                                       host_names=[host_name])
+    elif cluster_name:
+        objects = utils_common.get_clusters(service_instance, datacenter_name=datacenter_name, cluster_name=cluster_name)
+    elif datacenter_name:
+        objects = utils_common.get_datacenters(service_instance, datacenter_names=[datacenter_name])
+    else:
+        raise salt.exceptions.ArgumentValueError("_find_filtered_object requires at least one of datacenter_name, cluster_name and host_name")
+
+    if not objects:
+        return None
+    elif len(objects) == 1:
+        return objects[0]
+    else:
+        # this should be unreacheable and indicates a logic bug in this function
+        # the filters passed to this function should raise an exception if they could be ambiguous,
+        # signaling to the user that their usage is wrong during development, even if run against
+        # an environment with 1 host/cluster/datacenter.
+        raise Exception()
```

Comment:
I had thought about a NotImplementedError; adding a new exception is also an option.

I don't think it will ever happen, so I'm not sure how much effort to put in to it... but if it does happen it would be nice to get a report!